Add chapter for Lifetime Capture Rules 2024 (#316)
rust-lang/edition-guide: src/rust-2024/rpit-lifetime-capture.md

```rust,compile_fail
# #![feature(precise_capturing)]
fn capture<'a>(_: &'a ()) -> impl Sized + use<'a> {}

fn test<'a>(x: &'a ()) -> impl Sized + 'static {
    capture(x)
    //~^ ERROR lifetime may not live long enough
}
```

Conversely, this is OK:

```rust
# #![feature(precise_capturing)]
fn capture<'a>(_: &'a ()) -> impl Sized + use<> {}

fn test<'a>(x: &'a ()) -> impl Sized + 'static {
    capture(x) //~ OK
}
```

### Edition-specific rules when no `use<..>` bound is present

If the `use<..>` bound is not present, then the compiler uses edition-specific rules to decide which in-scope generic parameters to capture implicitly.

In all editions, all in-scope type and const generic parameters are captured implicitly when the `use<..>` bound is not present.  E.g.:

```rust
# #![feature(precise_capturing)]
fn f_implicit<T, const C: usize>() -> impl Sized {}
//                                    ~~~~~~~~~~
//                         No `use<..>` bound is present here.
//
// In all editions, the above is equivalent to:
fn f_explicit<T, const C: usize>() -> impl Sized + use<T, C> {}
```

> In Rust 2021 and earlier editions, when the `use<..>` bound is not present, generic lifetime parameters are only implicitly captured when they appear syntactically within a bound in RPIT opaque types in the signature of bare functions and associated functions and methods within inherent impls.  However, starting in Rust 2024, these in-scope generic lifetime parameters *are* implicitly captured.  E.g.:

Reviewer: ```suggestion
In Rust 2021 and earlier editions, when the `use<..>` bound is not present, generic lifetime parameters are only captured when they appear syntactically within a bound in RPIT opaque types in the signature of bare functions and associated functions and methods within inherent impls.  However, starting in Rust 2024, these in-scope generic lifetime parameters *are* unconditionally captured.  E.g.:
```

Author: Thanks, updated.

# RPIT lifetime capture rules

🚧 The 2024 Edition has not yet been released and hence this section is still "under construction".

This chapter describes changes related to the **Lifetime Capture Rules 2024** introduced in [RFC 3498], including opaque type *precise capturing* introduced in [RFC 3617].

[RFC 3498]: https://github.com/rust-lang/rfcs/pull/3498
[RFC 3617]: https://github.com/rust-lang/rfcs/pull/3617

## Summary

> - In all editions, `use<..>` bounds allow for capturing only the lifetime parameters needed in RPIT opaque types.

Reviewer: "needed" is awkward here.

Reviewer: I feel like this puts the cart before the horse. I'd probably describe the implcit behavior first, since that's why users are here in the first place, and then describe `use<>` as the convenient fix to all the capturing woes.

Author: I've remove the word and concept "needed" here, rearranged these bullet points, and removed one of the bullet points about `use<..>`.

Author: The other `use<..>`-related bullet point has now been removed also.

- If a `use<..>` bound is elided (i.e. not present) then the compiler automatically inserts one according to edition-specific rules.

> - In Rust 2024, *all* in-scope generic parameters, including lifetime parameters, are included in these automatically inserted `use<..>` bounds.

Reviewer: I know that you can get a lot of mileage out of explaining this in terms of `use<..>`, but it also feels obtuse, and possibly less approachable given that people are more familiar with just calling it capturing.

I would probably not frame this in terms of an implicit `use<..>`, because that's literally not what happens in the compiler. Let's just call it capturing?

Author: I feel like capturing isn't that widely understood, and that people are going to rapidly start thinking of this in a syntactic way in terms of `use<..>`, which is why it was framed this way, but I take the point, and I've reframed it in terms of capturing.

## Details

### Capturing and `use<..>` syntax

*Capturing* a generic parameter in an RPIT (return-position impl Trait) opaque type allows for that parameter to be used in the corresponding hidden type.  In Rust 1.82, we added `use<..>` bounds that allow specifying explicitly which generic parameters to capture.  E.g.:

```rust
# #![feature(precise_capturing)]
fn capture<'a, T>(x: &'a (), y: T) -> impl Sized + use<'a, T> {
    //                                ~~~~~~~~~~~~~~~~~~~~~~~
    //                             This is the RPIT opaque type.
    //
    //                                It captures `'a` and `T`.
    (x, y)
  //~~~~~~
  // The hidden type is: `(&'a (), T)`.
  //
  // This type can use `'a` and `T` because they were captured.
}
```

Currently, `use<..>` bounds must list all generic type and const parameters in scope, but they may exclude in-scope lifetime parameters.  E.g.:

```rust
# #![feature(precise_capturing)]
fn capture<'a, T>(_: &'a (), _: T) -> impl Sized + use<T> {
    //                                ~~~~~~~~~~~~~~~~~~~
    //                 The opaque type captures only `T` and not `'a`.
    ()
}
```

> You might want to capture more lifetimes in order to return a hidden type that uses those lifetimes.  Conversely, you might want to capture fewer lifetimes in order to allow the opaque type to be used in more places.  E.g., this is an error because the lifetime is captured despite the fact that the hidden type does not use the lifetime:

Reviewer: > You might want to capture more lifetimes [...]

Would be nice to show an example of how comes about on ed < 2024.

> You might want to capture fewer lifetimes [...]

Would be nice to show an example of how this happens on ed 2024.

Author: There are examples later in the document when we start talking about the migration.  Maybe we could go on a bit more there, but this chapter is already rather long, and at some point it feels like people should just read RFC 3498 if they want the full motivation.

Capturing is a subtle enough topic that the approach taken here is to introduce and it explain it first before getting into the edition-specific differences in the next section.

Author: This entire paragraph has now been removed from the document.

```rust,compile_fail
# #![feature(precise_capturing)]
fn capture<'a>(_: &'a ()) -> impl Sized + use<'a> {}

fn test<'a>(x: &'a ()) -> impl Sized + 'static {
    capture(x)
    //~^ ERROR lifetime may not live long enough
}
```

Conversely, this is OK since the opaque type does not capture the lifetime:

```rust
# #![feature(precise_capturing)]
fn capture<'a>(_: &'a ()) -> impl Sized + use<> {}

fn test<'a>(x: &'a ()) -> impl Sized + 'static {
    capture(x) //~ OK
}
```

> The bounds list of an RPIT opaque type may contain at most one `use<..>` bound.  If this bound is present, then the listed generic parameters are the only ones captured.

Reviewer: This feels misplaced. I would rather we move this to some reference section on `use`. Or really, we probably don't need to mention this caveat at all?

Author: I've removed this and the other bits that served to explain `use<..>` syntax.

### Edition-specific rules for elided `use<..>` bounds

> If the `use<..>` bound is *elided* (i.e., not present), then the compiler automatically inserts one according to edition-specific rules.

Reviewer: It is unnecessary to frame this in terms of an elided `use<..>` bound. It doesn't bring clarity IMO, over just saying "if `use<..>` is not present, then opaques capture parameters according to edition-specific rules".

Author: All uses of the word "elide" or "elided" and this framing have been removed.

> In all editions, all in-scope type and const generic parameters are added to this automatically-inserted `use<..>` bound.  E.g.:

Reviewer: ```suggestion
In all editions, all in-scope type and const generic parameters are captured.  E.g.:
```

Reviewer: e.g.

Author: All descriptions of automatically adding bounds have been removed.

```rust
# #![feature(precise_capturing)]
fn f_elided<T, const C: usize>() -> impl Sized {}
//                                  ~~~~~~~~~~
//                       The `use<..>` bound is elided here.
//
// In all editions, the above is equivalent to:
fn f_explicit<T, const C: usize>() -> impl Sized + use<T, C> {}
```

> In Rust 2021 and earlier editions, generic lifetime parameters are *not* included in this automatically-inserted `use<..>` bound for RPIT opaque types in the signature of bare functions and associated functions and methods within inherent impls.  However, starting in Rust 2024, these in-scope generic lifetime parameters *are* included.  E.g.:

Reviewer: So first, I think we can shorten this by just saying "are *not* always captured". Also, I may have missed it, but do we actually explain *when* lifetimes are captured pre-2024?

Author: This paragraph has been changed, as elsewhere, to remove discussion of automatically inserted `use<..>` bounds, and it now states how lifetime parameters are captured in Rust 2021 and earlier editions.

//   ~~~~~~~~~~~~~~~~~
// These generic parameters are in scope.
    fn f_elided<U>() -> impl Sized {}
    //          ~       ~~~~~~~~~~
    //          ^ This generic is in scope too.
    //                  ^
    //                  |
    //   The `use<..>` bound is elided here.
    //
    // In all editions, it's equivalent to:
    fn f_explicit<U>() -> impl Sized + use<T, U, C> {}
}
```

### Lifetimes from higher-ranked binders

Similarly, generic lifetime parameters introduced into scope by a higher-ranked `for<..>` binder are considered to be in scope.  E.g.:

```rust
# #![feature(precise_capturing)]
trait Tr<'a> { type Ty; }
impl Tr<'_> for () { type Ty = (); }

fn f_elided() -> impl for<'a> Tr<'a, Ty = impl Copy + use<>> {}
// In Rust 2021 and earlier, the above is equivalent to:
fn f_2021() -> impl for<'a> Tr<'a, Ty = impl Copy + use<>> {}
// In Rust 2024 and later, it's equivalent to:
//fn f_2024() -> impl for<'a> Tr<'a, Ty = impl Copy + use<'a>> {}
//                                        ~~~~~~~~~~~~~~~~~~~~
// However, note that the capturing of higher-ranked lifetimes in
// nested opaque types is not yet supported.
```

### Argument position impl Trait (APIT)

> Anonymous (i.e. unnamed) generic parameters created by the use of APIT (argument position impl Trait) are considered to be in scope when inserting an explicit `use<..>` bound for an elided one.  E.g.:

Reviewer: ```suggestion
Anonymous (i.e. unnamed) generic parameters created by the use of APIT (argument position impl Trait) are considered to be in scope for the purposes of capturing.  E.g.:
```

or something like that.

Author: All descriptions of automatically inserting `use<..>` bounds have been removed.
